pkg/bootstrapper/kubelet.go

```diff
@@ -0,0 +1,245 @@
+package bootstrapper
+
+import (
+	"fmt"
+	"time"
+
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+// kubeletService struct contains the kubelet specific service information
+type kubeletService struct {
+	// obj is a pointer to the Windows service object
+	obj *mgr.Service
+	// dependents contains a list of services dependent on the current service
+	dependents []*mgr.Service
+}
+
+// newKubeletService creates and returns a new kubeletService object
+func newKubeletService(ksvc *mgr.Service, svcDependents []*mgr.Service) (*kubeletService, error) {
+	if ksvc == nil {
+		return nil, fmt.Errorf("service object should not be nil")
+	}
+	return &kubeletService{
+		obj:        ksvc,
+		dependents: svcDependents,
+	}, nil
+}
+
+// getConfig retrieves service config from service object Config()
+func (k *kubeletService) getConfig() (mgr.Config, error) {
+	config, err := k.obj.Config()
+	if err != nil {
+		return mgr.Config{}, err
+	}
+	return config, nil
+}
+
+// start ensures that kubelet service is running
+func (k *kubeletService) start() error {
+	if k.obj == nil {
+		return fmt.Errorf("no kubelet service found")
+	}
+	// if the service is running we do not need to start it
+	isServiceRunning, err := k.isRunning()
+	if err != nil {
+		return fmt.Errorf("unable to check if Windows service is running: %v", err)
+	}
+	if isServiceRunning {
+		return nil
+	}
+	if err := k.obj.Start(); err != nil {
+		return err
+	}
+
+	if len(k.dependents) == 0 {
+		return nil
+	}
+	for _, dependent := range k.dependents {
+		err := startService(dependent)
+		if err != nil {
+			return fmt.Errorf("unable to start dependent service for %s", KubeletServiceName)
+		}
+	}
+	return nil
+}
+
+// control sends a signal to the service and waits until it changes state in response to the signal
+func (k *kubeletService) control(cmd svc.Cmd, desiredState svc.State) error {
+	status, err := k.obj.Control(cmd)
+	if err != nil {
+		return err
+	}
+	// Most of the rest of the function borrowed from https://godoc.org/golang.org/x/sys/windows/svc/mgr#Service.Control
+	timeout := time.Now().Add(serviceWaitTime)
+	for status.State != desiredState {
+		if timeout.Before(time.Now()) {
+			return fmt.Errorf("timeout waiting for service to go to state=%d", desiredState)
+		}
+		time.Sleep(300 * time.Millisecond)
+		status, err = k.obj.Query()
+		if err != nil {
+			return fmt.Errorf("could not retrieve service status: %v", err)
+		}
+	}
+	return nil
+}
+
+// stop ensures that the kubelet service is stopped
+func (k *kubeletService) stop() error {
+	isServiceRunning, err := k.isRunning()
+	if err != nil {
+		return fmt.Errorf("unable to check if kubelet service is running: %v", err)
+	}
+	if !isServiceRunning {
+		return nil
+	}
+	// the list of dependents is static here and contains one level of dependencies
```

Comment:
mentioning the approach here is not recursive/goes one level deep

Reply:
Please mention it in the `stop()` docs too. A reader of the docs should be able to see that.

Reply:
done